Automatic creation of a datasource in grafana

holgero opened the pull request:
This uses the /api/datasources endpoint of grafana to create a datasource automatically as soon as grafana becomes available. monit will retry this until it succeeds.

Commit:
Automatic creation of a datasource in grafana

This uses monit to retry until grafana becomes available.

holgero commented:
As I created the pull request, I started thinking if this better should be optional... OTOH grafana is pretty useless without at least one datasource.

What's your opinion?

Commit:
make datasource creation optional

holgero commented:
I made the automatic creation of the datasource optional, if no datasource.url is specified, nothing will be created.

vito commented:
Cool!

How easy would it be to make it idempotent, instead of a no-op if it's already configured? Also being able to configure multiple datasources would be nice. PR looks good as-implemented though, aside from a small typo in the monit config ("marker").

holgero commented:
Thanks for reviewing!

> How easy would it be to make it idempotent, instead of a no-op if it's already configured?

Yes, good point. I think I can do that with not too much effort.

> small typo in the monit config ("marker").

Hmm, this is probably just my bad English, not a typo. What would be a better name for the file that indicates that the script ran successful?

vito commented:
Ah my bad. I thought it was supposed to say "maker"; didn't realize it was a file check and not a process check.

Commit:
make datasource creation script idempotent

Update datasource, if it already exists instead of a noop.
Also improved logging.

holgero commented:
I didn't implement multiple datasources yet, but perhaps that could be another change that builds on this one?

vito commented:
Sounds good, thanks for making it idempotent!

vito commented on jobs/grafana/spec:
```diff
@@ -125,3 +126,26 @@ properties:
+
+  grafana.datasource.name:
+    description: "The name of the datasource that points to the metrics database"
+    default: influxdb
```
Could we kill all these defaults? I'd expect to have to specify basically all of them.

vito commented:
Trying this out now, if all is well I think I'll just kill the defaults myself and merge it in unless you feel strongly about them.

holgero commented:
That's fine by me.

On 10. November 2015 19:02:05 MEZ, Alex Suraci notifications@github.com wrote:

> Trying this out now, if all is well I think I'll just kill the defaults
> myself and merge it in unless you feel strongly about them.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/vito/grafana-boshrelease/pull/3#issuecomment-155515185

## 

Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

vito commented on jobs/grafana/templates/create_datasource.erb:
```diff
@@ -0,0 +1,40 @@
+<% if_p("grafana.datasource.url") do |datasource_url| %>
+exec &>> ${LOG_DIR}/datasource_create.log
+
+URL="http://<%= p("grafana.admin_username") %>:<%= p("grafana.admin_password") %>@localhost:<%= p("grafana.listen_port") %>/api/datasources"
```
Ran into a few problems with this:
1. My password has funny characters, so they'd need to be urlencoded. Maybe use the `-u` flag with `curl` instead; it seems to handle it properly.
2. Would prefer `127.0.0.1` over `localhost`.
3. My instance is configured with HTTPS, so this would need to go to that URI scheme instead (probably with `-k` to trust the local cert).

holgero replied:
I am sorry, I am at home already. But I can take care of the issues tomorrow.

Am 10. November 2015 19:12:59 MEZ, schrieb Alex Suraci notifications@github.com:

> > @@ -0,0 +1,40 @@
> > +#!/bin/bash
> > +
> > +LOG_DIR=/var/vcap/sys/log/grafana
> > +MARKERFILE=/var/vcap/sys/run/grafana/datasource_created
> > +
> > +<% if_p("grafana.datasource.url") do |datasource_url| %>
> > +exec &>> ${LOG_DIR}/datasource_create.log
> > +
> > +URL="http://<%= p("grafana.admin_username") %>:<%=
> > p("grafana.admin_password") %>@localhost:<%= p("grafana.listen_port")
> > %>/api/datasources"
> 
> Ran into a few problems with this:
> 1. My password has funny characters, so they'd need to be urlencoded.
>    Maybe use the `-u` flag with `curl` instead; it seems to handle it
>    properly.
> 2. Would prefer `127.0.0.1` over `localhost`.
> 3. My instance is configured with HTTPS, so this would need to go to
>    that URI scheme instead (probably with `-k` to trust the local cert).
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/vito/grafana-boshrelease/pull/3/files#r44442343

## 

Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

Commit:
Let curl handle user and passwords and...

...
* use 127.0.0.1 instead of localhost
* allow https scheme (and skip certificate checking)
* removed defaults in spec for datasource stuff

Signed-off-by: Jan Sievers <jan.sievers@sap.com>

holgero commented:
I hope we fixed the problems. (We didn't have a grafana running on https to test that part.)

holgero commented:
Hey Alex, did you have a chance to test it again yet?

vito commented:
Yep looks good. I'll try it again tomorrow. Thanks for the update!

vito commented:
Hmm, seeing something really weird with the updating, but I suspect it's a Grafana bug and not your fault. A few of the panes in my dashboard had their datasource switch to the default "grafana" datasource. I wonder if the `PUT` request isn't quite idempotent? :/

vito commented:
Going to merge this anyway and open an issue upstream if it keeps happening. Thanks!
